devel/fortio/cmd/grpcping/pingsrv.go

```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		fortio.Fatalf("failed to listen: %v", err)
+	}
+	grpcServer := grpc.NewServer()
+	reflection.Register(grpcServer)
+	healthServer := health.NewServer()
+	healthServer.SetServingStatus("ping", grpc_health_v1.HealthCheckResponse_SERVING)
+	grpc_health_v1.RegisterHealthServer(grpcServer, healthServer)
+	RegisterPingServerServer(grpcServer, &pingServer{})
+	fmt.Printf("Fortio %s grpc ping server listening on port %v\n", fortio.Version, port)
+	if err := grpcServer.Serve(socket); err != nil {
+		fortio.Fatalf("failed to start grpc server: %v", err)
+	}
+}
+
+func clientCall(serverAddr string, n int, payload string) {
+	conn, err := grpc.Dial(serverAddr, grpc.WithInsecure())
+	if err != nil {
+		fortio.Fatalf("failed to conect to %s: %v", serverAddr, err)
+	}
+	msg := &PingMessage{Payload: payload}
+	cli := NewPingServerClient(conn)
+	// Warm up:
+	_, err = cli.Ping(context.Background(), msg)
+	if err != nil {
+		fortio.Fatalf("grpc error from Ping0 %v", err)
+	}
+	skewHistogram := fortio.NewHistogram(-10, 2)
+	rttHistogram := fortio.NewHistogram(0, 10)
+	for i := 1; i <= n; i++ {
+		msg.Seq = int64(i)
+		t1a := time.Now().UnixNano()
+		msg.Ts = t1a
+		res1, err := cli.Ping(context.Background(), msg)
+		t2a := time.Now().UnixNano()
+		if err != nil {
+			fortio.Fatalf("grpc error from Ping1 %v", err)
+		}
+		t1b := res1.Ts
+		res2, err := cli.Ping(context.Background(), msg)
+		t3a := time.Now().UnixNano()
+		t2b := res2.Ts
+		if err != nil {
+			fortio.Fatalf("grpc error from Ping2 %v", err)
+		}
+		rt1 := t2a - t1a
+		rttHistogram.Record(float64(rt1) / 1000.)
+		rt2 := t3a - t2a
+		rttHistogram.Record(float64(rt2) / 1000.)
+		rtR := t2b - t1b
+		rttHistogram.Record(float64(rtR) / 1000.)
+		midR := t1b + (rtR / 2)
+		avgRtt := (rt1 + rt2 + rtR) / 3
+		x := (midR - t2a)
+		fortio.Infof("Ping RTT %d (avg of %d, %d, %d ns) clock skew %d",
+			avgRtt, rt1, rtR, rt2, x)
+		skewHistogram.Record(float64(x) / 1000.)
+		msg = res2
+	}
+	skewHistogram.Print(os.Stdout, "Clock skew histogram usec", 50)
+	rttHistogram.Print(os.Stdout, "RTT histogram usec", 50)
+}
+
+func healthCheck(serverAddr string, svcname string, n int) {
+	conn, err := grpc.Dial(serverAddr, grpc.WithInsecure())
+	if err != nil {
+		fortio.Fatalf("failed to conect to %s: %v", serverAddr, err)
+	}
+	msg := &grpc_health_v1.HealthCheckRequest{Service: svcname}
+	cli := grpc_health_v1.NewHealthClient(conn)
+	rttHistogram := fortio.NewHistogram(0, 10)
+	statuses := make(map[grpc_health_v1.HealthCheckResponse_ServingStatus]int64)
+
+	for i := 1; i <= n; i++ {
+		start := time.Now()
+		res1, err := cli.Check(context.Background(), msg)
+		dur := time.Since(start)
+		if err != nil {
+			fortio.Fatalf("grpc error from Check %v", err)
+		}
+		statuses[res1.Status]++
+		rttHistogram.Record(dur.Seconds() * 1000000.)
+	}
+	rttHistogram.Print(os.Stdout, "RTT histogram usec", 50)
+	fmt.Printf("Statuses %v\n", statuses)
+}
+
+func main() {
+	flag.Parse()
+	if *hostFlag != "" {
```

Comment:
in the future, you may want to consider https://github.com/spf13/cobra to allow subcommands that may be better suited for this purpose (instead of `grpcping -host ...`, you would have something like `grpcping server ...` which _might_ be easier to grok (and allow file separation of client/server code)).